validator/valnode/redis/consumer.go

```diff
@@ -93,17 +108,11 @@ func (s *ValidationServer) Start(ctx_in context.Context) {
 					// There's nothing in the queue.
 					return time.Second
 				}
-				valRun := s.spawner.Launch(req.Value, moduleRoot)
-				res, err := valRun.Await(ctx)
-				if err != nil {
-					log.Error("Error validating", "request value", req.Value, "error", err)
-					return 0
-				}
-				if err := c.SetResult(ctx, req.ID, res); err != nil {
-					log.Error("Error setting result for request", "id", req.ID, "result", res, "error", err)
-					return 0
+				select {
+				case <-ctx.Done():
+				case workQueue <- workUnit{req, moduleRoot}:
```

Comment:
If only one worker is configured, then this is sort of always buffered. This goroutine is always waiting with a new work unit ready to go while the background goroutine is working on the previous one. I think we could fix this by making a makeshift semaphore with a buffered channel. Something like:

```go
type requestToken struct {}

// once at startup
requestTokens := make(channel requestToken, bufferSize + workers)
for _ := range cap(requestTokens) {
    requestTokens <- requestToken{}
}

// in the consumer goroutine
// TODO context cancelation listening via select
token := <-requestTokens
req, err := c.Consume(ctx)
workQueue <- workUnit{req, moduleRoot, token}

// in the worker
work := <-workQueue
// to use a defer we'd want to move doing a work item to its own function, and that's probably a good refactor
defer func() { requestTokens <- work.token }()
// [do the work]
```